src/java.base/share/classes/java/lang/module/Resolver.java

```diff
@@ -649,6 +606,77 @@ private Map<ResolvedModule, Set<ResolvedModule>> makeGraph(Configuration cf) {
             }
         } while (changed);
 
+        // Handle automatic modules in a separate step
+
+        // automatic modules read all selected modules and all modules
+        // in parent configurations
+        for (Map.Entry<ResolvedModule, Set<ResolvedModule>> entry : g1.entrySet()) {
+            ResolvedModule m1 = entry.getKey();
+
+            if (!m1.descriptor().isAutomatic()) {
+                continue;
+            }
+
+            String name = m1.name();
+            Set<ResolvedModule> m1Reads = entry.getValue();
+
+            // reads all selected modules
+            for (ModuleReference mref2 : nameToReference.values()) {
+                ModuleDescriptor descriptor2 = mref2.descriptor();
+                String name2 = descriptor2.name();
+
+                if (!name.equals(name2)) {
+                    ResolvedModule m2 = nameToResolved.get(name2);
+                    assert m2 != null;
+                    m1Reads.add(m2);
+                }
+            }
+
+            // reads all modules in parent configurations
+            for (Configuration parent : parents) {
+                parent.configurations()
+                        .map(Configuration::modules)
+                        .flatMap(Set::stream)
+                        .forEach(m1Reads::add);
+            }
+        }
+
+        // m2 is automatic module => m1 reads all automatic modules that m2 reads
+        for (Map.Entry<ResolvedModule, Set<ResolvedModule>> entry : g1.entrySet()) {
+            ResolvedModule m1 = entry.getKey();
+
+            if (m1.descriptor().isAutomatic()) {
+                continue;
+            }
+
+            Set<ResolvedModule> m1Reads = entry.getValue();
+
+            for (ResolvedModule m2 : m1Reads) {
+                if (!m2.descriptor().isAutomatic()) {
+                    continue;
+                }
+
+                Set<ResolvedModule> m2Reads;
+                if (m2.configuration() == cf) {
+                    // this configuration
+                    m2Reads = g1.get(m2);
+                } else {
+                    // parent configuration, already resolved
+                    // TODO: does this allocate a copy of the set every time?
```

Comment:
It doesn't. `List.copyOf` and `Set.copyOf` returns the passed immutable collection instance if input is already an immutable collection.

Reply:
The input did not seem immutable however